### PR TITLE
Strict type `Dependabot::Docker::Requirement::CredentialsFinder`

### PR DESCRIPTION
**Pull request by raj-meka**

### CODE REVIEW

File: docker/lib/dependabot/docker/utils/credentials_finder.rb [hunk 1]
```diff
@@ -76,7 +76,7 @@ def build_aws_credentials(registry_details)
 
           # Otherwise, we need to use the provided Access Key ID and secret to
           # generate a temporary username and password
-          @authorization_tokens ||= {}
+          @authorization_tokens ||= T.let({}, T.nilable(T::Hash[T.untyped, T.untyped]))
```
Comment (abdulapopoola) [inline docker/lib/dependabot/docker/utils/credentials_finder.rb:79]: can this be typed some more?

File: docker/lib/dependabot/docker/utils/credentials_finder.rb [hunk 2]
```diff
@@ -32,14 +33,13 @@ def credentials_for_registry(registry_hostname)
...
         end
 
+        sig { returns(String) }
         def base_registry
-          @base_registry ||= credentials.find do |cred|
-            cred["type"] == "docker_registry" && cred.replaces_base?
-          end
-          @base_registry ||= { "registry" => DEFAULT_DOCKER_HUB_REGISTRY, "credentials" => nil }
-          @base_registry["registry"]
+          @base_registry ||= T.let(nil, T.nilable(Dependabot::Credential))
```
Comment (abdulapopoola) [inline docker/lib/dependabot/docker/utils/credentials_finder.rb:38]: why are we changing the code?

File: docker/lib/dependabot/docker/utils/helpers.rb
Comment (JamieMagee) [inline docker/lib/dependabot/docker/utils/helpers.rb:1]: This file is also in #9890

File: docker/lib/dependabot/docker/utils/helpers.rb [hunk 3]
```diff
@@ -1,14 +0,0 @@
-# typed: true
-# frozen_string_literal: true
-
```
[the remark this answers is not in the document]
Comment (abdulapopoola) [reply docker/lib/dependabot/docker/utils/helpers.rb:3]: Can you merge main into this branch? This file was already fixed yesterday

File: docker/lib/dependabot/docker/utils/credentials_finder.rb [hunk 4]
```diff
@@ -32,14 +34,16 @@ def credentials_for_registry(registry_hostname)
...
 
+        sig { returns(T.nilable(String)) }
         def base_registry
+          @base_registry ||= T.let(nil, T.nilable(String))
           @base_registry ||= credentials.find do |cred|
             cred["type"] == "docker_registry" && cred.replaces_base?
-          end
-          @base_registry ||= { "registry" => DEFAULT_DOCKER_HUB_REGISTRY, "credentials" => nil }
-          @base_registry["registry"]
+          end&.fetch("registry", nil)
```
Comment (abdulapopoola) [inline docker/lib/dependabot/docker/utils/credentials_finder.rb:42]: why is this included?

File: docker/lib/dependabot/docker/utils/credentials_finder.rb [hunk 5]
```diff
@@ -32,14 +34,18 @@ def credentials_for_registry(registry_hostname)
           build_aws_credentials(registry_details)
         end
 
+        sig { returns(T.nilable(String)) }
         def base_registry
+          @base_registry ||= T.let(nil, T.nilable(Dependabot::Credential))
           @base_registry ||= credentials.find do |cred|
             cred["type"] == "docker_registry" && cred.replaces_base?
           end
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/utils/credentials_finder.rb:42]: You can do this in a single statement:

```suggestion
          @base_registry ||= T.let(
          credentials.find do |cred|
            cred["type"] == "docker_registry" && cred.replaces_base?
          end,
          T.nilable(Dependabot::Credential)
          )
```

You'll need to fix the indentation though. The suggestion above is syntactically correct, but Rubocop will complain.

**Review by JamieMagee — APPROVED** [review]

**Review by abdulapopoola — APPROVED** [review]
